### ISSUE TITLE
Training on VOC some questions about mAP and tensorboard usage

### ISSUE THREAD

**Issue by achel-x** [eriklindernoren/PyTorch-YOLOv3#750]
## I perform the training on the VOC dataset,
COCO is very time-consumming. So I change the dataset to VOC2012. I trained 300 epochs. Here is the record by tensorboard.

![image](https://user-images.githubusercontent.com/67953243/143551886-d6a3ce70-031d-40af-9a32-ae4842a9d6a6.png)

The current mAP is around 48. I am not sure what the mAP can be obtained in VOC. 
In my view, I thought increase the epochs or reducing the learning rate in a certain stage maybe bring some improvement. 
But

## Breakpoint trainig need change cfg?
I observed the learning rate plays an important role in training. 
And I can use the code to train with my the checkpoint in the 300th epoch. But when I want to continue training, how should I adapt the cfg file?（learning rate and steps setting is questioned)
Moreover, every time I start the training there is a new tensorboard file log generated. I want to continue record the training process in the same tensorboard log, which continuing plot along the previos curve. But I have no idea how to solve it.
Please give me some advice. Thanks in advance.

Looking forward to instruction. Thanks in advance.

**Comment by achel-x** [comment]
I use my model to run detect. Unfortunatly, the image detected is not good. There are many objects haven't been detected. 
![image](https://user-images.githubusercontent.com/67953243/143588556-fe202acc-a592-435c-a1df-e44fc03b56e5.png)
This is the output image. The bird haven't been detected. 
Some small obect can be detected. 
![image](https://user-images.githubusercontent.com/67953243/143590634-84645b0a-060c-4dce-b332-6ee38e77f3af.png)

I found 3 threshold in code **iou_thres, conf_thres and nms_thres**. But the 3 values differ in cfg, detect.py .
In yolov3.cfg there is **ignore_thresh** and **truth_thres**
In train.py : iou_thres = 0.5, conf_thres = 0.1 and the nms_thres = 0.5
While in detect.py: conf_thres=0.5, num_thres=0.4 where different from train.py.
I try to revise the setting in detect.py. But modifying the two values has no effect on the detection results.

I can find some in other blogs to know about nms_thres and ignore_thres. 
The nms_thres is used to decrease the num of predicted bbox. Lager value means more bbox will be kept.
The ignore_thres compare the predicted bbox with ground truth. Does the abnormal phoenomenon correlated to this value?

While the **iou_thres, conf_thres and truth_tres**  confuse me.  I hope to get some comments on these values.

I'm sorry I have so many puzzles. I hope you don't mind and enlighten me.

**Comment by cjr13018** [comment]
> ## I perform the training on the VOC dataset,
> COCO is very time-consumming. So I change the dataset to VOC2012. I trained 300 epochs. Here is the record by tensorboard.
> 
> ![image](https://user-images.githubusercontent.com/67953243/143551886-d6a3ce70-031d-40af-9a32-ae4842a9d6a6.png)
> 
> The current mAP is around 48. I am not sure what the mAP can be obtained in VOC. In my view, I thought increase the epochs or reducing the learning rate in a certain stage maybe bring some improvement. But
> 
> ## Breakpoint trainig need change cfg?
> I observed the learning rate plays an important role in training. And I can use the code to train with my the checkpoint in the 300th epoch. But when I want to continue training, how should I adapt the cfg file?（learning rate and steps setting is questioned) Moreover, every time I start the training there is a new tensorboard file log generated. I want to continue record the training process in the same tensorboard log, which continuing plot along the previos curve. But I have no idea how to solve it. Please give me some advice. Thanks in advance.
> 
> Looking forward to instruction. Thanks in advance.
I also want to use VOC  dataset,but I do not know hao to   do.

**Comment by Flova** [comment]
> The current mAP is around 48. I am not sure what the mAP can be obtained in VOC. In my view, I thought increase the epochs or reducing the learning rate in a certain stage maybe bring some improvement. But

I am not sure about the possible mAP in VOC2012 ether. There is definitely some hyperparameter optimization involved if you want to get very high scores.

The learning rate is reduced during training as described in the `.cfg` file.
![image](https://user-images.githubusercontent.com/15075613/145026713-4d49d240-63ff-4644-a7a7-3dea39495bf5.png)

The steps are optimizer steps (one is done after each batch (not mini-batch)). And the scales define the factor the learning rate should be multiplied at this point.

The effects of the learning rate reduction can be clearly seen in the mAP graph when you compare it with the learning rate graph:
![image](https://user-images.githubusercontent.com/15075613/145027029-daea8547-8d69-49bc-af76-0ba5b35d2bf3.png)
![image](https://user-images.githubusercontent.com/15075613/145027060-258a95b6-faeb-497f-bf51-5aed355a8198.png)

> ## Breakpoint trainig need change cfg?
> 
> I observed the learning rate plays an important role in training. And I can use the code to train with my the checkpoint in the 300th epoch. But when I want to continue training, how should I adapt the cfg file?（learning rate and steps setting is questioned) Moreover, every time I start the training there is a new tensorboard file log generated. I want to continue record the training process in the same tensorboard log, which continuing plot along the previos curve. But I have no idea how to solve it. 

You can adjust the settings in the cfg and schedule different learning rates. I recommend this strongly instead of changing the cfg manually during training and continuing the training with a checkpoint. Appending different tensor board trials to each other should be no problem them. You could never the less change this to wall time to see all your experiments sequentially:
![image](https://user-images.githubusercontent.com/15075613/145027479-2bc1917e-a907-4968-a496-17410b8587db.png)

**Comment by Flova** [comment]
> I use my model to run detect. Unfortunatly, the image detected is not good. There are many objects haven't been detected. ![image](https://user-images.githubusercontent.com/67953243/143588556-fe202acc-a592-435c-a1df-e44fc03b56e5.png) This is the output image. The bird hasn't been detected. Some small obect can be detected. ![image](https://user-images.githubusercontent.com/67953243/143590634-84645b0a-060c-4dce-b332-6ee38e77f3af.png)

Are you sure that your data loading process for VOC 2012 is correct and that the labels and images are loaded correctly?
Did you use any pre-trained encoder weights? They can boot the performance quite a bit.
The described mAP is not perfect ether, so it can be expected that some object are not detected.
Are the images from the test set?

> I found 3 threshold in code **iou_thres, conf_thres and nms_thres**. But the 3 values differ in cfg, detect.py . In yolov3.cfg there is **ignore_thresh** and **truth_thres** In train.py : iou_thres = 0.5, conf_thres = 0.1 and the nms_thres = 0.5 While in detect.py: conf_thres=0.5, num_thres=0.4 where different from train.py. I try to revise the setting in detect.py. But modifying the two values has no effect on the detection results.

The nms is only relevant if you have multiple detections overlapping. And the conf_thres allows you to increase the sensitivity. So if you lower this value you should get more (but worse) detections. The conf thresh in the train and test should not be touched, but in detect you can lower is and you should see more boxes.

> I can find some in other blogs to know about nms_thres and ignore_thres. The nms_thres is used to decrease the num of predicted bbox. Lager value means more bbox will be kept. The ignore_thres compare the predicted bbox with ground truth. Does the abnormal phoenomenon correlated to this value?
> 
> While the **iou_thres, conf_thres and truth_tres** confuse me. I hope to get some comments on these values.
> 
> I'm sorry I have so many puzzles. I hope you don't mind and enlighten me.

ignore_thres is a training hyperparameter and can be left untouched most of the time.

**Comment by achel-x** [comment]
@cjr13018 
First, convert the Annotations .xml file to yolo format( .txt). Here are many scripts you can search in the Internet. 
eg https://github.com/AlexeyAB/darknet/blob/master/scripts/voc_label.py
But how to use them, maybe you need some time to search others blog for understand.
Then, follow the tutorial in the homepage in this repo **train on cutsom data,** create 'custom.data' , put the images and labels in corresponding path. 
Third, change the cfg **the conv** just before yolo the filter is (num_class + 5)*3, （ 3 places altogther correspond to 3 yolo head). VOC is 75(coco is 255 originally) and num_class. 
You may need to refer some blogs for specific operations. Indeed, I can also write one, but I am busy with other aspects. Although my work needs to use Yolo, it focuses on other fields( That's why I am not familiar with the principle of yolo)

**Comment by achel-x** [comment]
@Flova Thank you for your explanantions. I am sure the way to load the VOC dataset correctly. Your code is very concise and easy to understand, which helps me quickly grasp the structure of Yolo code. While after several attempts, I haven't got the ideal results. I used others' repo and the expected values appeared. 
Anyway, thank you for your work, which provides a good guide for my study of Yolo. 
In addition, according to my study of another code, I found that some data augmentation methods do have a great effect on map.
